File: api/v1/schemas/portfolio.py
# -*- coding: utf-8 -*-
"""Portfolio API schemas."""

from __future__ import annotations

from datetime import date, datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator
from pydantic_core import PydanticCustomError


PortfolioInstrumentType = Literal["equity", "etf", "qdii", "adr_ads", "daily_leveraged_product", "unknown"]
PortfolioVerificationStatus = Literal["verified", "provisional", "missing"]


class PortfolioInstrumentCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(..., min_length=1, max_length=32)
    market: Literal["cn", "hk", "us", "jp", "kr", "tw"]
    quote_currency: str = Field(..., min_length=3, max_length=8)
    instrument_type: PortfolioInstrumentType
    underlying_symbol: Optional[str] = Field(None, max_length=32)
    underlying_market: Optional[Literal["cn", "hk", "us", "jp", "kr", "tw"]] = None
    underlying_currency: Optional[str] = Field(None, max_length=8)
    leverage_factor: Optional[float] = Field(None, gt=0)
    daily_reset: bool = False
    conversion_ratio: Optional[float] = Field(None, gt=0)
    trade_lot_size: float = Field(..., gt=0)
    requires_premium_check: bool = False
    verification_status: PortfolioVerificationStatus = "missing"
    evidence_source: Optional[str] = Field(None, max_length=512)
    evidence_as_of: Optional[datetime] = Field(
        None,
        description="Evidence timestamp with an explicit timezone offset; stored as UTC.",
    )
    metadata: Optional[Dict[str, Any]] = None


class PortfolioInstrumentUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    quote_currency: Optional[str] = Field(None, min_length=3, max_length=8)
    instrument_type: Optional[PortfolioInstrumentType] = None
    underlying_symbol: Optional[str] = Field(None, max_length=32)
    underlying_market: Optional[Literal["cn", "hk", "us", "jp", "kr", "tw"]] = None
    underlying_currency: Optional[str] = Field(None, max_length=8)
    leverage_factor: Optional[float] = Field(None, gt=0)
    daily_reset: Optional[bool] = None
    conversion_ratio: Optional[float] = Field(None, gt=0)
    trade_lot_size: Optional[float] = Field(None, gt=0)
    requires_premium_check: Optional[bool] = None
    verification_status: Optional[PortfolioVerificationStatus] = None
    evidence_source: Optional[str] = Field(None, max_length=512)
    evidence_as_of: Optional[datetime] = Field(
        None,
        description="Evidence timestamp with an explicit timezone offset; stored as UTC.",
    )
    metadata: Optional[Dict[str, Any]] = None


class PortfolioInstrumentItem(BaseModel):
    id: int
    symbol: str
    market: str
    quote_currency: str
    instrument_type: str
    underlying_symbol: Optional[str] = None
    underlying_market: Optional[str] = None
    underlying_currency: Optional[str] = None
    leverage_factor: Optional[float] = None
    daily_reset: bool
    conversion_ratio: Optional[float] = None
    trade_lot_size: float
    requires_premium_check: bool
    verification_status: str
    evidence_source: Optional[str] = None
    evidence_as_of: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PortfolioInstrumentListResponse(BaseModel):
    items: List[PortfolioInstrumentItem] = Field(default_factory=list)


class PortfolioRiskPolicyUpsertRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    min_cash_buffer_pct: Optional[float] = Field(None, ge=0, le=100)
    max_single_position_pct: Optional[float] = Field(None, ge=0, le=100)
    max_sector_pct: Optional[float] = Field(None, ge=0, le=100)
    max_high_risk_product_pct: Optional[float] = Field(None, ge=0, le=100)
    max_portfolio_drawdown_pct: Optional[float] = Field(None, ge=0, le=100)


class PortfolioRiskPolicyItem(BaseModel):
    id: int
    min_cash_buffer_pct: float
    max_single_position_pct: float
    max_sector_pct: float
    max_high_risk_product_pct: float
    max_portfolio_drawdown_pct: float
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PortfolioRiskPolicyResponse(BaseModel):
    policy: Optional[PortfolioRiskPolicyItem] = None


class PortfolioAccountCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=64)
    broker: Optional[str] = Field(None, max_length=64)
    market: Literal["cn", "hk", "us", "jp", "kr", "tw"] = "cn"
    base_currency: str = Field("CNY", min_length=3, max_length=8)
    owner_id: Optional[str] = Field(None, max_length=64)


class PortfolioAccountUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=64)
    broker: Optional[str] = Field(None, max_length=64)
    market: Optional[Literal["cn", "hk", "us", "jp", "kr", "tw"]] = None
    base_currency: Optional[str] = Field(None, min_length=3, max_length=8)
    owner_id: Optional[str] = Field(None, max_length=64)
    is_active: Optional[bool] = None


class PortfolioAccountItem(BaseModel):
    id: int
    owner_id: Optional[str] = None
    name: str
    broker: Optional[str] = None
    market: str
    base_currency: str
    is_active: bool
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PortfolioAccountListResponse(BaseModel):
    accounts: List[PortfolioAccountItem] = Field(default_factory=list)


class PortfolioTradeCreateRequest(BaseModel):
    account_id: int
    symbol: str = Field(..., min_length=1, max_length=16)
    trade_date: date
    side: Literal["buy", "sell"]
    quantity: float = Field(..., gt=0)
    price: float = Field(..., gt=0)
    fee: float = Field(0.0, ge=0)
    tax: float = Field(0.0, ge=0)
    market: Optional[Literal["cn", "hk", "us", "jp", "kr", "tw"]] = None
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    trade_uid: Optional[str] = Field(None, max_length=128)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioCashLedgerCreateRequest(BaseModel):
    account_id: int
    event_date: date
    direction: Literal["in", "out"]
    amount: float = Field(..., gt=0)
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioCorporateActionCreateRequest(BaseModel):
    account_id: int
    symbol: str = Field(..., min_length=1, max_length=16)
    effective_date: date
    action_type: Literal["cash_dividend", "split_adjustment"]
    market: Optional[Literal["cn", "hk", "us", "jp", "kr", "tw"]] = None
    currency: Optional[str] = Field(None, min_length=3, max_length=8)
    cash_dividend_per_share: Optional[float] = Field(None, ge=0)
    split_ratio: Optional[float] = Field(None, gt=0)
    note: Optional[str] = Field(None, max_length=255)


class PortfolioEventCreatedResponse(BaseModel):
    id: int


class PortfolioDeleteResponse(BaseModel):
    deleted: int


class PortfolioTradeListItem(BaseModel):
    id: int
    account_id: int
    trade_uid: Optional[str] = None
    symbol: str
    market: str
    currency: str
    trade_date: str
    side: str
    quantity: float
    price: float
    fee: float
    tax: float
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioTradeListResponse(BaseModel):
    items: List[PortfolioTradeListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioCashLedgerListItem(BaseModel):
    id: int
    account_id: int
    event_date: str
    direction: str
    amount: float
    currency: str
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioCashLedgerListResponse(BaseModel):
    items: List[PortfolioCashLedgerListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioCorporateActionListItem(BaseModel):
    id: int
    account_id: int
    symbol: str
    market: str
    currency: str
    effective_date: str
    action_type: str
    cash_dividend_per_share: Optional[float] = None
    split_ratio: Optional[float] = None
    note: Optional[str] = None
    created_at: Optional[str] = None


class PortfolioCorporateActionListResponse(BaseModel):
    items: List[PortfolioCorporateActionListItem] = Field(default_factory=list)
    total: int
    page: int
    page_size: int


class PortfolioPositionItem(BaseModel):
    symbol: str
    market: str
    currency: str
    quantity: float
    avg_cost: float
    total_cost: float
    last_price: float
    market_value_base: float
    unrealized_pnl_base: float
    unrealized_pnl_pct: Optional[float] = None
    valuation_currency: str
    price_source: str = "unknown"
    price_provider: Optional[str] = None
    price_date: Optional[str] = None
    price_stale: bool = False
    price_available: bool = True
    data_quality: str = "ok"
    limitations: List[str] = Field(default_factory=list)


class PortfolioPositionAnalysisRequest(BaseModel):
    account_id: Optional[int] = Field(None, description="Optional account id; required when a symbol is held in multiple accounts")
    analysis_phase: Literal["auto", "premarket", "intraday", "postmarket"] = "auto"
    force: bool = Field(False, description="Force refresh analysis inputs without bypassing duplicate in-flight tasks")
    research_snapshot_hash: Optional[str] = Field(
        None,
        min_length=64,
        max_length=64,
        pattern=r"^[0-9a-fA-F]{64}$",
        description="Expected hash of the preflight portfolio research snapshot",
    )
    research_cutoff: Optional[datetime] = Field(
        None,
        description="Timezone-aware cutoff of the preflight portfolio research snapshot",
    )

    @model_validator(mode="after")
    def validate_research_snapshot_binding(self) -> "PortfolioPositionAnalysisRequest":
        if (self.research_snapshot_hash is None) != (self.research_cutoff is None):
            raise PydanticCustomError(
                "research_snapshot_binding_incomplete",
                "research_snapshot_hash and research_cutoff must be provided together",
            )
        if self.research_cutoff is not None:
            if self.research_cutoff.tzinfo is None or self.research_cutoff.utcoffset() is None:
                raise PydanticCustomError(
                    "research_cutoff_timezone_missing",
                    "research_cutoff must include an explicit timezone offset",
                )
        return self


class PortfolioAccountSnapshot(BaseModel):
    account_id: int
    account_name: str
    owner_id: Optional[str] = None
    broker: Optional[str] = None
    market: str
    base_currency: str
    as_of: str
    cost_method: str
    total_cash: float
    total_market_value: float
    total_equity: float
    realized_pnl: float
    unrealized_pnl: float
    fee_total: float
    tax_total: float
    fx_stale: bool
    data_quality: str = "ok"
    limitations: List[str] = Field(default_factory=list)
    positions: List[PortfolioPositionItem] = Field(default_factory=list)


class PortfolioSnapshotResponse(BaseModel):
    as_of: str
    cost_method: str
    currency: str
    account_count: int
    total_cash: float
    total_market_value: float
    total_equity: float
    realized_pnl: float
    unrealized_pnl: float
    fee_total: float
    tax_total: float
    fx_stale: bool
    data_quality: str = "ok"
    limitations: List[str] = Field(default_factory=list)
    accounts: List[PortfolioAccountSnapshot] = Field(default_factory=list)


class PortfolioAnalysisRuntimeProof(BaseModel):
    architecture: Literal["single", "multi"]
    automatic_multi_agent: bool


class PortfolioResearchSnapshotResponse(BaseModel):
    schema_version: str
    cutoff: str
    timezone: str
    cost_method: str
    analysis_runtime: PortfolioAnalysisRuntimeProof
    universe_hash: str
    snapshot_hash: str
    accounts: List[Dict[str, Any]] = Field(default_factory=list)
    positions: List[Dict[str, Any]] = Field(default_factory=list)
    instruments: List[Dict[str, Any]] = Field(default_factory=list)
    benchmarks: List[Dict[str, Any]] = Field(default_factory=list)
    risk_policy: Optional[Dict[str, Any]] = None
    risk_budget: Optional[Dict[str, Any]] = None
    hard_blockers: List[Dict[str, Any]] = Field(default_factory=list)
    limitations: List[str] = Field(default_factory=list)
    completeness: str


class PortfolioResearchBaselineRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    research_snapshot_hash: str = Field(
        ...,
        min_length=64,
        max_length=64,
        pattern=r"^[0-9a-fA-F]{64}$",
    )
    research_cutoff: datetime

    @model_validator(mode="after")
    def validate_research_cutoff(self) -> "PortfolioResearchBaselineRequest":
        if self.research_cutoff.tzinfo is None or self.research_cutoff.utcoffset() is None:
            raise PydanticCustomError(
                "research_cutoff_timezone_missing",
                "research_cutoff must include an explicit timezone offset",
            )
        return self


class PortfolioResearchBaselineItem(BaseModel):
    account_id: int
    account_name: Optional[str] = None
    market: str
    symbol: str
    name: Optional[str] = None
    display_label: str
    selection_key: str
    currency: Optional[str] = None
    quantity: Optional[float] = None
    instrument_type: str = "unknown"
    quote: Dict[str, Any] = Field(default_factory=dict)
    history: Dict[str, Any] = Field(default_factory=dict)
    trend: Optional[Dict[str, Any]] = None
    current_signal_id: Optional[int] = None
    position_action: Literal["hold", "reduce", "exit"]
    incremental_action: Literal["add_in_batches", "wait", "no_add"]
    core_reason: Optional[str] = None
    hard_blockers: List[str] = Field(default_factory=list)
    risk_flags: List[Dict[str, Any]] = Field(default_factory=list)
    exception_reasons: List[str] = Field(default_factory=list)
    evidence_status: str
    research_level: Literal["baseline"] = "baseline"
    detail_recommended: bool = False
    sizing_allowed: bool = False


class PortfolioResearchBaselineCandidate(BaseModel):
    selection_key: str
    display_label: str
    market: str
    symbol: str
    account_ids: List[int] = Field(default_factory=list)
    reasons: List[str] = Field(default_factory=list)
    priority: int
    recommended: bool


class PortfolioResearchBaselineResponse(BaseModel):
    schema_version: Literal["portfolio-research-baseline-v1"]
    snapshot_hash: str
    cutoff: str
    market_data_cutoff: str
    ledger_position_count: int
    baseline_row_count: int
    coverage_reconciled: bool
    portfolio_risk_flags: List[Dict[str, Any]] = Field(default_factory=list)
    items: List[PortfolioResearchBaselineItem] = Field(default_factory=list)
    suggested_deep_analysis: List[PortfolioResearchBaselineCandidate] = Field(default_factory=list)
    deep_analysis_started: bool = False


class PortfolioImportTradeItem(BaseModel):
    trade_date: str
    symbol: str
    side: Literal["buy", "sell"]
    quantity: float
    price: float
    fee: float
    tax: float
    trade_uid: Optional[str] = None
    dedup_hash: str
    currency: Optional[str] = None


class PortfolioImportParseResponse(BaseModel):
    broker: str
    record_count: int
    skipped_count: int
    error_count: int
    records: List[PortfolioImportTradeItem] = Field(default_factory=list)
    errors: List[str] = Field(default_factory=list)


class PortfolioImportCommitResponse(BaseModel):
    account_id: int
    record_count: int
    inserted_count: int
    duplicate_count: int
    failed_count: int
    dry_run: bool
    errors: List[str] = Field(default_factory=list)


class PortfolioImportBrokerItem(BaseModel):
    broker: str
    aliases: List[str] = Field(default_factory=list)
    display_name: Optional[str] = None


class PortfolioImportBrokerListResponse(BaseModel):
    brokers: List[PortfolioImportBrokerItem] = Field(default_factory=list)


class PortfolioFxRefreshResponse(BaseModel):
    as_of: str
    account_count: int
    refresh_enabled: bool
    disabled_reason: Optional[str] = None
    pair_count: int
    updated_count: int
    stale_count: int
    error_count: int


class PortfolioDecisionSignalRiskItem(BaseModel):
    account_id: Optional[int] = None
    symbol: str
    market: str
    signal: Dict[str, Any] = Field(default_factory=dict)


class PortfolioDecisionSignalRiskBlock(BaseModel):
    available: bool = True
    total: int = 0
    actions: Dict[str, int] = Field(default_factory=dict)
    items: List[PortfolioDecisionSignalRiskItem] = Field(default_factory=list)


class PortfolioRiskResponse(BaseModel):
    as_of: str
    account_id: Optional[int] = None
    cost_method: str
    currency: str
    thresholds: Dict[str, Any] = Field(default_factory=dict)
    concentration: Dict[str, Any] = Field(default_factory=dict)
    sector_concentration: Dict[str, Any] = Field(default_factory=dict)
    drawdown: Dict[str, Any] = Field(default_factory=dict)
    stop_loss: Dict[str, Any] = Field(default_factory=dict)
    decision_signal_risk: PortfolioDecisionSignalRiskBlock = Field(default_factory=PortfolioDecisionSignalRiskBlock)

File: apps/dsa-web/src/types/portfolio.ts
import type { DecisionSignalItem } from './decisionSignals';

export type PortfolioCostMethod = 'fifo' | 'avg';
export type PortfolioSide = 'buy' | 'sell';
export type PortfolioCashDirection = 'in' | 'out';
export type PortfolioCorporateActionType = 'cash_dividend' | 'split_adjustment';
export type PortfolioInstrumentType = 'equity' | 'etf' | 'qdii' | 'adr_ads' | 'daily_leveraged_product' | 'unknown';
export type PortfolioVerificationStatus = 'verified' | 'provisional' | 'missing';

export interface PortfolioInstrumentItem {
  id: number;
  symbol: string;
  market: 'cn' | 'hk' | 'us' | 'jp' | 'kr' | 'tw';
  quoteCurrency: string;
  instrumentType: PortfolioInstrumentType;
  underlyingSymbol?: string | null;
  underlyingMarket?: 'cn' | 'hk' | 'us' | 'jp' | 'kr' | 'tw' | null;
  underlyingCurrency?: string | null;
  leverageFactor?: number | null;
  dailyReset: boolean;
  conversionRatio?: number | null;
  tradeLotSize: number;
  requiresPremiumCheck: boolean;
  verificationStatus: PortfolioVerificationStatus;
  evidenceSource?: string | null;
  evidenceAsOf?: string | null;
  metadata: Record<string, unknown>;
}

export type PortfolioInstrumentInput = Omit<PortfolioInstrumentItem, 'id' | 'metadata'> & {
  metadata?: Record<string, unknown>;
};

export interface PortfolioRiskPolicyItem {
  id: number;
  minCashBufferPct: number;
  maxSinglePositionPct: number;
  maxSectorPct: number;
  maxHighRiskProductPct: number;
  maxPortfolioDrawdownPct: number;
}

export type PortfolioRiskPolicyInput = Omit<PortfolioRiskPolicyItem, 'id'>;

export interface PortfolioPointInTimeEligibility {
  scope: 'current_prospective';
  prospectiveDecisionEligible: boolean;
  historicalReplayEligible: false;
  sourceCutoffs: Record<string, string | null>;
  blockers: string[];
}

export interface PortfolioFrozenDecisionSignal {
  id: number;
  market: string;
  stockCode: string;
  stockName?: string | null;
  reason?: string | null;
  status: string;
  createdAt?: string | null;
  updatedAt?: string | null;
  metadata: Record<string, unknown>;
}

export interface PortfolioResearchSnapshotResponse {
  snapshotHash: string;
  executionIdentityHash: string;
  cutoff: string;
  scope: PortfolioResearchScopeItem[];
  scopeHash?: string | null;
  completeness: string;
  positions: Array<Record<string, unknown>>;
  instruments: Array<Record<string, unknown>>;
  pointInTime: PortfolioPointInTimeEligibility;
  decisionSignals: PortfolioFrozenDecisionSignal[];
  hardBlockers: Array<{ code: string; scope: string; symbol?: string; market?: string }>;
  limitations: string[];
}

export interface PortfolioResearchScopeItem {
  accountId: number;
  market: 'cn' | 'hk' | 'us' | 'jp' | 'kr' | 'tw';
  symbol: string;
}

export interface PortfolioResearchEvidenceItem {
  accountId: number;
  symbol: string;
  market: string;
  currency: string;
  benchmarkCode?: string | null;
  status: 'ready' | 'insufficient';
  price?: Record<string, unknown> | null;
  benchmark?: Record<string, unknown> | null;
  fx?: Record<string, unknown> | null;
  productEvidence?: Record<string, unknown> | null;
  blockers: string[];
}

export interface PortfolioResearchEvidencePrepareResponse {
  schemaVersion: 'portfolio-research-evidence-prepare-v2';
  scope: PortfolioResearchScopeItem[];
  preparedAt: string;
  cutoff: string;
  asOf: string;
  status: 'ready' | 'partial' | 'empty';
  positionCount: number;
  readyCount: number;
  insufficientCount: number;
  items: PortfolioResearchEvidenceItem[];
}

export interface PortfolioResearchBaselineRequest {
  researchSnapshotHash: string;
  researchCutoff: string;
  researchScope: PortfolioResearchScopeItem[];
}

export interface PortfolioResearchExecutionCheckRequest {
  researchSnapshotHash: string;
  researchExecutionIdentityHash: string;
  researchCutoff: string;
  researchScope: PortfolioResearchScopeItem[];
}

export interface PortfolioResearchExecutionCheckItem {
  accountId: number;
  market: string;
  symbol: string;
  name?: string | null;
  status: 'ready' | 'insufficient';
  referenceEvidence: Record<string, unknown>;
  currentEvidence: Record<string, unknown>;
  changedFields: string[];
  blockers: string[];
  requiresReconfirmation: boolean;
}

export interface PortfolioResearchExecutionCheckResponse {
  schemaVersion: 'portfolio-research-execution-check-v1';
  checkedAt: string;
  researchSnapshotHash: string;
  scope: PortfolioResearchScopeItem[];
  status: 'ready' | 'partial';
  requiresReconfirmation: boolean;
  items: PortfolioResearchExecutionCheckItem[];
}

export type PortfolioUserInstruction = 'add' | 'hold' | 'reduce' | 'exit' | 'insufficient';

export interface PortfolioResearchBaselineItem {
  accountId: number;
  accountName?: string | null;
  market: string;
  symbol: string;
  name?: string | null;
  displayLabel: string;
  selectionKey: string;
  currency?: string | null;
  quantity?: number | null;
  instrumentType: string;
  quote: Record<string, unknown>;
  history: Record<string, unknown>;
  trend?: Record<string, unknown> | null;
  currentSignalId?: number | null;
  positionAction: 'hold' | 'reduce' | 'exit';
  incrementalAction: 'add_in_batches' | 'wait' | 'no_add';
  userInstruction: PortfolioUserInstruction;
  coreReason?: string | null;
  hardBlockers: string[];
  riskFlags: Array<Record<string, unknown>>;
  exceptionReasons: string[];
  evidenceStatus: string;
  researchLevel: 'baseline';
  detailRecommended: boolean;
  sizingAllowed: boolean;
}

export interface PortfolioResearchBaselineCandidate {
  selectionKey: string;
  displayLabel: string;
  market: string;
  symbol: string;
  accountIds: number[];
  reasons: string[];
  priority: number;
  recommended: boolean;
}

export interface PortfolioResearchBaselineResponse {
  schemaVersion: 'portfolio-research-baseline-v1';
  snapshotHash: string;
  cutoff: string;
  marketDataCutoff: string;
  ledgerPositionCount: number;
  baselineRowCount: number;
  coverageReconciled: boolean;
  portfolioRiskFlags: Array<Record<string, unknown>>;
  items: PortfolioResearchBaselineItem[];
  suggestedDeepAnalysis: PortfolioResearchBaselineCandidate[];
  deepAnalysisStarted: boolean;
}

export interface PortfolioAccountItem {
  id: number;
  ownerId?: string | null;
  name: string;
  broker?: string | null;
  market: 'cn' | 'hk' | 'us' | 'jp' | 'kr' | 'tw';
  baseCurrency: string;
  isActive: boolean;
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface PortfolioAccountListResponse {
  accounts: PortfolioAccountItem[];
}

export interface PortfolioAccountCreateRequest {
  name: string;
  broker?: string;
  market: 'cn' | 'hk' | 'us' | 'jp' | 'kr' | 'tw';
  baseCurrency: string;
  ownerId?: string;
}

export interface PortfolioPositionItem {
  symbol: string;
  market: string;
  currency: string;
  quantity: number;
  avgCost: number;
  totalCost: number;
  lastPrice: number;
  marketValueBase: number;
  unrealizedPnlBase: number;
  unrealizedPnlPct?: number | null;
  valuationCurrency: string;
  priceSource?: 'realtime_quote' | 'history_close' | 'missing' | string;
  priceProvider?: string | null;
  priceDate?: string | null;
  priceStale?: boolean;
  priceAvailable?: boolean;
  dataQuality?: 'ok' | 'partial' | string;
  limitations?: string[];
}

type PortfolioPositionAnalysisOptions = {
  accountId?: number;
  analysisPhase?: 'auto' | 'premarket' | 'intraday' | 'postmarket';
  force?: boolean;
};

type PortfolioResearchBinding =
  | {
      researchSnapshotHash: string;
      researchCutoff: string;
      researchScope: PortfolioResearchScopeItem[];
    }
  | {
      researchSnapshotHash?: never;
      researchCutoff?: never;
      researchScope?: never;
    };

export type PortfolioPositionAnalysisRequest = PortfolioPositionAnalysisOptions & PortfolioResearchBinding;

export interface PortfolioAccountSnapshot {
  accountId: number;
  accountName: string;
  ownerId?: string | null;
  broker?: string | null;
  market: string;
  baseCurrency: string;
  asOf: string;
  costMethod: PortfolioCostMethod;
  totalCash: number;
  totalMarketValue: number;
  totalEquity: number;
  realizedPnl: number;
  unrealizedPnl: number;
  feeTotal: number;
  taxTotal: number;
  fxStale: boolean;
  dataQuality?: 'ok' | 'partial' | string;
  limitations?: string[];
  positions: PortfolioPositionItem[];
}

export interface PortfolioSnapshotResponse {
  asOf: string;
  costMethod: PortfolioCostMethod;
  currency: string;
  accountCount: number;
  totalCash: number;
  totalMarketValue: number;
  totalEquity: number;
  realizedPnl: number;
  unrealizedPnl: number;
  feeTotal: number;
  taxTotal: number;
  fxStale: boolean;
  dataQuality?: 'ok' | 'partial' | string;
  limitations?: string[];
  accounts: PortfolioAccountSnapshot[];
}

export interface PortfolioConcentrationItem {
  symbol: string;
  marketValueBase: number;
  weightPct: number;
  isAlert: boolean;
}

export interface PortfolioSectorConcentrationItem {
  sector: string;
  marketValueBase: number;
  weightPct: number;
  symbolCount: number;
  isAlert: boolean;
}

export interface PortfolioDrawdownBlock {
  available?: boolean;
  seriesPoints: number;
  maxDrawdownPct: number | null;
  currentDrawdownPct: number | null;
  alert: boolean;
  fxStale: boolean;
  limitations?: string[];
}

export interface PortfolioStopLossItem {
  accountId: number;
  symbol: string;
  avgCost: number;
  lastPrice: number;
  lossPct: number;
  nearThresholdPct: number;
  isTriggered: boolean;
}

export interface PortfolioDecisionSignalRiskItem {
  accountId?: number | null;
  symbol: string;
  market: string;
  signal: Partial<DecisionSignalItem>;
}

export interface PortfolioDecisionSignalRiskBlock {
  available: boolean;
  total: number;
  actions: {
    sell?: number;
    reduce?: number;
    alert?: number;
    [key: string]: number | undefined;
  };
  items: PortfolioDecisionSignalRiskItem[];
}

export interface PortfolioRiskResponse {
  asOf: string;
  accountId?: number | null;
  costMethod: PortfolioCostMethod;
  currency: string;
  thresholds: Record<string, number>;
  concentration: {
    totalMarketValue: number;
    topWeightPct: number;
    alert: boolean;
    topPositions: PortfolioConcentrationItem[];
  };
  sectorConcentration: {
    totalMarketValue: number;
    topWeightPct: number;
    alert: boolean;
    topSectors: PortfolioSectorConcentrationItem[];
    coverage: Record<string, number>;
    errors: string[];
  };
  drawdown: PortfolioDrawdownBlock;
  stopLoss: {
    nearAlert: boolean;
    triggeredCount: number;
    nearCount: number;
    items: PortfolioStopLossItem[];
  };
  decisionSignalRisk?: PortfolioDecisionSignalRiskBlock;
}

export interface PortfolioTradeCreateRequest {
  accountId: number;
  symbol: string;
  tradeDate: string;
  side: PortfolioSide;
  quantity: number;
  price: number;
  fee?: number;
  tax?: number;
  market?: 'cn' | 'hk' | 'us' | 'jp' | 'kr' | 'tw';
  currency?: string;
  tradeUid?: string;
  note?: string;
}

export interface PortfolioCashLedgerCreateRequest {
  accountId: number;
  eventDate: string;
  direction: PortfolioCashDirection;
  amount: number;
  currency?: string;
  note?: string;
}

export interface PortfolioCorporateActionCreateRequest {
  accountId: number;
  symbol: string;
  effectiveDate: string;
  actionType: PortfolioCorporateActionType;
  market?: 'cn' | 'hk' | 'us' | 'jp' | 'kr' | 'tw';
  currency?: string;
  cashDividendPerShare?: number;
  splitRatio?: number;
  note?: string;
}

export interface PortfolioEventCreatedResponse {
  id: number;
}

export interface PortfolioDeleteResponse {
  deleted: number;
}

export interface PortfolioTradeListItem {
  id: number;
  accountId: number;
  tradeUid?: string | null;
  symbol: string;
  market: string;
  currency: string;
  tradeDate: string;
  side: PortfolioSide;
  quantity: number;
  price: number;
  fee: number;
  tax: number;
  note?: string | null;
  createdAt?: string | null;
}

export interface PortfolioTradeListResponse {
  items: PortfolioTradeListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface PortfolioCashLedgerListItem {
  id: number;
  accountId: number;
  eventDate: string;
  direction: PortfolioCashDirection;
  amount: number;
  currency: string;
  note?: string | null;
  createdAt?: string | null;
}

export interface PortfolioCashLedgerListResponse {
  items: PortfolioCashLedgerListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface PortfolioCorporateActionListItem {
  id: number;
  accountId: number;
  symbol: string;
  market: string;
  currency: string;
  effectiveDate: string;
  actionType: PortfolioCorporateActionType;
  cashDividendPerShare?: number | null;
  splitRatio?: number | null;
  note?: string | null;
  createdAt?: string | null;
}

export interface PortfolioCorporateActionListResponse {
  items: PortfolioCorporateActionListItem[];
  total: number;
  page: number;
  pageSize: number;
}

export interface PortfolioImportTradeItem {
  tradeDate: string;
  symbol: string;
  side: PortfolioSide;
  quantity: number;
  price: number;
  fee: number;
  tax: number;
  tradeUid?: string | null;
  dedupHash: string;
  currency?: string | null;
}

export interface PortfolioImportParseResponse {
  broker: string;
  recordCount: number;
  skippedCount: number;
  errorCount: number;
  records: PortfolioImportTradeItem[];
  errors: string[];
}

export interface PortfolioImportCommitResponse {
  accountId: number;
  recordCount: number;
  insertedCount: number;
  duplicateCount: number;
  failedCount: number;
  dryRun: boolean;
  errors: string[];
}

export interface PortfolioImportBrokerItem {
  broker: string;
  aliases: string[];
  displayName?: string;
}

export interface PortfolioImportBrokerListResponse {
  brokers: PortfolioImportBrokerItem[];
}

export interface PortfolioFxRefreshResponse {
  asOf: string;
  accountCount: number;
  refreshEnabled?: boolean;
  disabledReason?: string | null;
  pairCount: number;
  updatedCount: number;
  staleCount: number;
  errorCount: number;
}
